doc/sphinx/proof-engine/ltac.rst

```diff
@@ -936,38 +1192,80 @@ We can perform pattern matching on goals using the following expression:
+      Hypotheses are matched from the last hypothesis (which is by default the newest
+      hypothesis) to the first until the :tacn:`apply` succeeds.
 
-      Using lazymatch instead of match will perform the same pattern matching
-      procedure but will commit to the first matching branch with the first
-      matching combination of hypotheses rather than trying a new matching if
-      the right-hand side fails. If the right-hand side of the selected branch
-      is a tactic with backtracking points, then subsequent failures cause
-      this tactic to backtrack.
+      .. coqtop:: reset all
+
+         Goal forall A B : Prop, A -> B -> (A->B).
+         intros.
+         match goal with
+         | H : _ |- _ => idtac "apply " H; apply H
+         end.
+
+.. _match_goal_hyps_rev_ex:
+
+   .. example:: Matching hypotheses with reverse
+
+      Hypotheses are matched from the first hypothesis to the last until the :tacn:`apply` succeeds.
+
+      .. coqtop:: reset all
+
+         Goal forall A B : Prop, A -> B -> (A->B).
+         intros.
+         match reverse goal with
+         | H : _ |- _ => idtac "apply " H; apply H
+         end.
+
+.. _match_goal_multiple_hyps_ex:
+
+   .. example:: Multiple ways to match hypotheses
+
+      Every possible match for the hypotheses is evaluated until the right-hand
+      side succeeds.  Note that `H1` and `H2` are never matched to the same hypothesis.
+      Observe that the number of permutations can grow as the factorial
+      of the number of hypotheses and hypothesis patterns.
+
+      .. coqtop:: reset all
+
+         Goal forall A B : Prop, A -> B -> (A->B).
+         intros A B H.
+         match goal with
+         | H1 : _, H2 : _ |- _ => idtac "match " H1 H2; fail
+         | _ => idtac
+         end.
+
+   .. todo need examples for:
+      match_context_rule ::= [ {*, @match_hyp } |- @match_pattern ] => @ltac_expr
+      match_hyp ::= | @name := {? [ @match_pattern ] : } @match_pattern
+
+.. todo PR The following items (up to numgoals) are part of "value_tactic".  I'd like to make
+   this a subsection and explain that they all return values.  How do I get a 5th-level section title?
 
 Filling a term context
 ~~~~~~~~~~~~~~~~~~~~~~
 
 The following expression is not a tactic in the sense that it does not
 produce subgoals but generates a term to be used in tactic expressions:
 
-.. tacn:: context @ident [@ltac_expr]
+.. tacn:: context @ident [ @term ]
 
-   :n:`@ident` must denote a context variable bound by a context pattern of a
+   .. todo @term is actually @lconstr (term with holes), which is spliced away
+      Seems like it's worth restoring that distinction
+
+   Replaces the context variable
+   :n:`@ident` bound by a context pattern of a
    match expression. This expression evaluates replaces the hole of the
    value of :n:`@ident` by the value of :n:`@ltac_expr`.
 
+   :tacn:`context` is a :token:`value_tactic`.
+
    .. exn:: Not a context variable.
       :undocumented:
```

Comment:
No, it's not redundant, but I think they should perhaps be merged (or else this bit should absolutely be linked to from the explanation in the `match` section), as you need this form to understand what the `@ident` does in matching on a context.  In particular, the example in the match section is an example of both forms.

Reply:
@JasonGross OK, IIUC you need to use them together to make a substitution, as in this contrived example that I'll add for the `context` tactic.  But seems better to keep them separate but linked in the doc:

```
Goal True /\ True.

match goal with
| |- context G [True] => let x := context G [False] in idtac x
end.

(* (False /\ True) *)
```

Reply:
> OK, IIUC you need to use them together to make a substitution

This is correct.  You can use the `match` form without the substitution form, but not the other way around (and the `ident` between `context` and `[...]` in the `match` form is useless without the substitution form).

If you want a less contrived example, here's a useful tactic from fiat-crypto:
```coq
Ltac split_in_context_by ident funl funr tac :=
  repeat match goal with
         | [ H : context p [ident] |- _ ] =>
           let H0 := context p[funl] in
           let H1 := context p[funr] in
           let H0' := (eval cbv beta in H0) in
           let H1' := (eval cbv beta in H1) in
           assert H0' by (tac H);
           assert H1' by (tac H);
           clear H
         end.
Ltac split_in_context ident funl funr :=
  split_in_context_by ident funl funr ltac:(fun H => apply H).

Ltac split_iff := split_in_context iff (fun a b : Prop => a -> b) (fun a b : Prop => b -> a).

Ltac split_contravariant_or := split_in_context_by or (fun A B : Prop => A) (fun A B : Prop => B) ltac:(fun H => intros; eauto 100 using H, or_introl, or_intror, ex_intro, exist, existT with nocore).

Ltac split_and := split_in_context and (fun a b : Type => a) (fun a b : Type => b).
Ltac split_prod := split_in_context prod (fun a b : Type => a) (fun a b : Type => b).
```

It lets you do nifty things like
```coq
Goal forall f g : nat -> Prop, (forall x, f x <-> g x) -> f 0 -> g 0.
  intros.
  (* 1 subgoal (ID 6)

  f, g : nat -> Prop
  H : forall x : nat, f x <-> g x
  H0 : f 0
  ============================
  g 0
   *)
  auto. (* no-op *)
  split_iff.
  (* 1 subgoal (ID 17)

  f, g : nat -> Prop
  H0 : f 0
  H1 : forall x : nat, f x -> g x
  H2 : forall x : nat, g x -> f x
  ============================
  g 0
   *)
  auto.
Qed.
```

Reply:
An interesting example, thanks.  I added a todo on the `context` tactic linking here.  I think it would be more appropriate in a future section or chapter that covers Ltac techniques.  It would need a bit of explanation.